src/python/pants/util/process_handler.py

```diff
@@ -27,6 +27,19 @@
 subprocess = subprocess3
 
 
+class Popen(subprocess3.Popen):
+  def __init__(self, args, **kwargs):
+    # This should not be necessary, but in the daemon context, it is - otherwise any modifications
+    # to os.environ are not reflected in the child process's environment.
+    # I have no idea why.
```

Comment:
discussed on slack, but for posterity:

without an explicit env dict, `execv` is used [here](https://github.com/google/python-subprocess32/blob/master/_posixsubprocess.c#L523-L526) and the docs for that say `The environment for the new process image is taken from the environ variable of the current process image`
where `environ` there is `extern char **environ` at the libc level. so the fundamental problem here is that we're successfully setting the heremetic environment in the daemon case at the /python/ level but not at the /c/ level because of the current approach to replacing `os.environ` outright.

I cut a quick fix for that at #5898 which also makes the integration test here pass.